Public/Src/Cache/ContentStore/Distributed/Redis/RedisDatabaseAdapter.cs

```diff
@@ -286,16 +293,65 @@ public async Task<BoolResult> ExecuteBatchOperationAsync(Context context, IRedis
                         },
                         cancellationToken);
                     await batch.NotifyConsumersOfSuccess();
+
+                    RedisOperationSucceeded();
                     return BoolResult.Success;
                 }
                 catch (Exception ex)
                 {
                     batch.NotifyConsumersOfFailure(ex);
+                    OnRedisException(context, ex);
                     return new BoolResult(ex);
                 }
             }
         }
 
+        private void RedisOperationSucceeded()
+        {
+            // If the operation is successful, then reset the error count to 0.
+            Interlocked.Exchange(ref _connectionErrorCount, 0);
+        }
+
+        private void OnRedisException(Context context, Exception exception)
+        {
+            if (IsRedisConnectionException(exception))
+            {
+                if (Interlocked.Increment(ref _connectionErrorCount) >= _connectionErrorCountLimit)
+                {
+                    lock (_resetConnectionsLock)
+                    {
+                        if (_connectionErrorCount >= _connectionErrorCountLimit)
+                        {
+                            context.Warning($"Reset redis connection due to connectivity issues. ConnectionErrorCount={_connectionErrorCountLimit}.");
+                            Interlocked.Exchange(ref _connectionErrorCount, 0);
+                            _databaseFactory.ResetConnectionMultiplexer();
+                        }
+                    }
+                }
+            }
+        }
+
+        private void HandleRedisException(Exception exception)

```

Comment:
Unclear to me when HandleRedisException is the code path vs when OnRedisException is the code path. Should we only have one? If we need both, shouldn't they both route to same logic (e.g. both could trigger reset?)